// @rei-network/database/dist/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.bloomBitsKey = exports.txLookupKey = exports.receiptsKey = exports.hashToNumberKey = exports.numberToHashKey = exports.bodyKey = exports.headerKey = exports.tdKey = exports.bufBE8 = exports.BLOOM_BITS_PREFIX = exports.TX_LOOKUP_PREFIX = exports.RECEIPTS_PREFIX = exports.BLOOM_BITS_SECTION_COUNT = exports.CLIQUE_BLOCK_SIGNERS_KEY = exports.CLIQUE_VOTES_KEY = exports.CLIQUE_SIGNERS_KEY = exports.HEAD_BLOCK_KEY = exports.HEAD_HEADER_KEY = exports.HEADS_KEY = void 0;
// Geth compatible DB keys
const HEADS_KEY = 'heads';
exports.HEADS_KEY = HEADS_KEY;
/**
 * Current canonical head for light sync
 */
const HEAD_HEADER_KEY = 'LastHeader';
exports.HEAD_HEADER_KEY = HEAD_HEADER_KEY;
/**
 * Current canonical head for full sync
 */
const HEAD_BLOCK_KEY = 'LastBlock';
exports.HEAD_BLOCK_KEY = HEAD_BLOCK_KEY;
/**
 * Cique signers
 */
const CLIQUE_SIGNERS_KEY = 'CliqueSigners';
exports.CLIQUE_SIGNERS_KEY = CLIQUE_SIGNERS_KEY;
/**
 * Clique votes
 */
const CLIQUE_VOTES_KEY = 'CliqueVotes';
exports.CLIQUE_VOTES_KEY = CLIQUE_VOTES_KEY;
/**
 * Cique block signers (snapshot)
 */
const CLIQUE_BLOCK_SIGNERS_KEY = 'CliqueBlockSignersSnapshot';
exports.CLIQUE_BLOCK_SIGNERS_KEY = CLIQUE_BLOCK_SIGNERS_KEY;
/**
 * Bloom bits section count
 */
const BLOOM_BITS_SECTION_COUNT = 'scount';
exports.BLOOM_BITS_SECTION_COUNT = BLOOM_BITS_SECTION_COUNT;
/**
 * headerPrefix + number + hash -> header
 */
const HEADER_PREFIX = Buffer.from('h');
/**
 * headerPrefix + number + hash + tdSuffix -> td
 */
const TD_SUFFIX = Buffer.from('t');
/**
 * headerPrefix + number + numSuffix -> hash
 */
const NUM_SUFFIX = Buffer.from('n');
/**
 * blockHashPrefix + hash -> number
 */
const BLOCK_HASH_PEFIX = Buffer.from('H');
/**
 * bodyPrefix + number + hash -> block body
 */
const BODY_PREFIX = Buffer.from('b');
/**
 * receiptPrfix
 */
const RECEIPTS_PREFIX = Buffer.from('r');
exports.RECEIPTS_PREFIX = RECEIPTS_PREFIX;
/**
 * txLookupPrefix
 */
const TX_LOOKUP_PREFIX = Buffer.from('l');
exports.TX_LOOKUP_PREFIX = TX_LOOKUP_PREFIX;
/**
 * bloomBitsIndexPrefix
 */
const BLOOM_BITS_PREFIX = Buffer.from('B');
exports.BLOOM_BITS_PREFIX = BLOOM_BITS_PREFIX;
// Utility functions
/**
 * Convert BN to big endian Buffer
 */
const bufBE8 = (n) => n.toArrayLike(Buffer, 'be', 8);
exports.bufBE8 = bufBE8;
const tdKey = (n, hash) => Buffer.concat([HEADER_PREFIX, bufBE8(n), hash, TD_SUFFIX]);
exports.tdKey = tdKey;
const headerKey = (n, hash) => Buffer.concat([HEADER_PREFIX, bufBE8(n), hash]);
exports.headerKey = headerKey;
const bodyKey = (n, hash) => Buffer.concat([BODY_PREFIX, bufBE8(n), hash]);
exports.bodyKey = bodyKey;
const numberToHashKey = (n) => Buffer.concat([HEADER_PREFIX, bufBE8(n), NUM_SUFFIX]);
exports.numberToHashKey = numberToHashKey;
const hashToNumberKey = (hash) => Buffer.concat([BLOCK_HASH_PEFIX, hash]);
exports.hashToNumberKey = hashToNumberKey;
const receiptsKey = (n, hash) => Buffer.concat([RECEIPTS_PREFIX, bufBE8(n), hash]);
exports.receiptsKey = receiptsKey;
const txLookupKey = (hash) => Buffer.concat([TX_LOOKUP_PREFIX, hash]);
exports.txLookupKey = txLookupKey;
const bloomBitsKey = (bit, section, hash) => {
    const buf = Buffer.alloc(10);
    buf.writeUInt16BE(bit);
    buf.writeBigUInt64BE(BigInt(section.toString()), 2);
    return Buffer.concat([BLOOM_BITS_PREFIX, buf, hash]);
};
exports.bloomBitsKey = bloomBitsKey;
