// @rei-network/structure/dist/receipt/extension.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LogExtension = exports.ReceiptExtension = void 0;
const ethereumjs_util_1 = require("ethereumjs-util");
class ReceiptExtension {
    constructor(block, tx, gasUsed, txIndex) {
        var _a;
        this.blockHash = block.hash();
        this.blockNumber = block.header.number;
        this.from = tx.getSenderAddress().toBuffer();
        this.contractAddress = tx.to ? undefined : (0, ethereumjs_util_1.generateAddress)(this.from, tx.nonce.toArrayLike(Buffer));
        this.gasUsed = gasUsed;
        this.to = (_a = tx === null || tx === void 0 ? void 0 : tx.to) === null || _a === void 0 ? void 0 : _a.toBuffer();
        this.transactionHash = tx.hash();
        this.transactionIndex = txIndex;
    }
}
exports.ReceiptExtension = ReceiptExtension;
class LogExtension {
    constructor(receipt, logIndex) {
        this.blockHash = receipt.blockHash;
        this.blockNumber = receipt.blockNumber;
        this.transactionHash = receipt.transactionHash;
        this.transactionIndex = receipt.transactionIndex;
        this.logIndex = logIndex;
    }
}
exports.LogExtension = LogExtension;
